src/OpenColorIO/ops/fixedfunction/FixedFunctionOpCPU.cpp

```diff
@@ -653,6 +690,145 @@ void Renderer_ACES_DarkToDim10_Fwd::apply(const void * inImg, void * outImg, lon
     }
 }
 
+float compress(float dist, float thr, float scale, float power, float one_over_power)
+{
+    // Normalize distance outside threshold by scale factor
+    const float nd = (dist - thr) / scale;
+    const float p = powf(nd, power);
+
+    // Compress
+    return thr + scale * nd / (powf(1.0f + p, one_over_power));
+}
+
+float uncompress(float dist, float thr, float scale, float power, float one_over_power)
+{
+    // Avoid singularity
+    if (dist >= (thr + scale))
+    {
+        return dist;
+    }
+    else
+    {
+        // Normalize distance outside threshold by scale factor
+        const float nd = (dist - thr) / scale;
+        const float p = powf(nd, power);
+
+        return thr + scale * powf(-(p / (p - 1.0f)), one_over_power);
+    }
+}
+
+template <typename Func>
+float gamut_comp(float val, float ach, float lim, float thr, float scale, float power, float one_over_power, Func f)
+{
+    // Disable compression
+    if (ach == 0.0f)
+    {
+        return 0.0f;
+    }
+
+    // Disable compression, avoid nan
+    if (lim < 1.0001f)
+    {
+        return val;
+    }
+
+    // Distance from the achromatic axis, aka inverse RGB ratios
+    const float dist = (ach - val) / std::fabs(ach);
+
+    // No compression below threshold
+    if (dist < thr)
+    {
+        return val;
+    }
+
+    // Compress / Uncompress distance with parameterized shaper function
+    const float comprDist = f(dist, thr, scale, power, one_over_power);
```

Comment:
How will you do a function pointer on the GPU?  (There is no requirement that the CPU and GPU code are exactly the same of course, but it does help the readability and maintenance slightly.)  Also, does this prevent declaring compress an inline function?  Given that it's in the pixel loop, it might be best to avoid an actual function call.

Perhaps it would be clearer to split your gamut_comp function into separate forward and inverse versions and this would avoid the need for the template and the function pointer?  And then the GPU could look the same as well.  Just a suggestion, I'm fine if you prefer to leave it as is.

Reply:
The GPU code path doesn't currently rely on this templated function argument but should be fairly close, I guess I followed the split approach you mentioned at a slight cost in code duplication.

But the rest probably belongs to a follow optimisation PR, again, I could be wrong but I would avoid making assumption without profiling, compiler should be pretty good about inlining and I don't believe the `__inline` used elsewhere in the code actually provide any guarantee the compiler will do such a thing. But only profiling will tell :)